src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/ProxyService.scala

```diff
@@ -75,13 +75,35 @@ class ProxyService(proxyConfig: ProxyConfig,
     }
   }
 
+  private[leonardo] val clusterInternalIdCache = CacheBuilder.newBuilder()
+    .expireAfterWrite(proxyConfig.cacheExpiryTime.toMinutes, TimeUnit.MINUTES)
+    .maximumSize(proxyConfig.cacheMaxSize)
+    .build(
+      new CacheLoader[(GoogleProject, ClusterName), Future[ClusterInternalId]] {
+        def load(key: (GoogleProject, ClusterName)) = {
+          val (googleProject, clusterName) = key
+          dbRef.inTransaction { dataAccess =>
+            dataAccess.clusterQuery.getActiveClusterInternalIdByName(googleProject, clusterName)
+          } map {
+            case Some(id) => id
+            case None => throw ProxyException(googleProject, clusterName)
+          }
+        }
+      }
+    )
+
+  def getCachedClusterInternalId(googleProject: GoogleProject, clusterName: ClusterName): Future[ClusterInternalId] = {
+    clusterInternalIdCache.get((googleProject, clusterName))
```

Comment:
should this return `Future[Option[ClusterInterald]]`?

Reply:
You know, that might be better -- we can deal with the `None` case outside of the cache, in `authCheck`. Will make that change.